File: backend/models/strategy.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Index, Integer, String, func, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.core.database import Base

if TYPE_CHECKING:
    from backend.models.driver import Driver
    from backend.models.race import Session


class StrategyPrediction(Base):
    __tablename__ = "strategy_predictions"
    # Backs alert_service._latest_undercut_scores: filters by session_id,
    # groups by driver_id, aggregates MAX(predicted_at) — then re-joins on
    # (driver_id, predicted_at) to fetch the winning row. DESC matches the
    # "most recent prediction" access pattern; the single-column indexes
    # below on session_id/driver_id don't help SQLAlchemy plan the composite
    # filter+group+max as one index scan.
    __table_args__ = (
        Index(
            "ix_strategy_predictions_session_driver_predicted_at",
            "session_id",
            "driver_id",
            text("predicted_at DESC"),
        ),
        # Backs strategy_service.get_strategy_prediction_history: filters by
        # session_id + driver_id, orders by lap_number ASC — a distinct access
        # pattern from the DESC-by-predicted_at index above (progression-over-
        # time view vs. "most recent" lookup), so it needs its own index rather
        # than reusing that one.
        Index(
            "ix_strategy_predictions_session_driver_lap_number",
            "session_id",
            "driver_id",
            "lap_number",
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    # No standalone index=True here: ix_strategy_predictions_session_driver_predicted_at
    # above already leads with session_id, so a separate single-column index
    # on it would be a pure duplicate.
    session_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sessions.id", ondelete="CASCADE"),
        nullable=False,
    )
    driver_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("drivers.id"), nullable=False, index=True
    )
    predicted_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    # Nullable: added 2026-08-26 (Day 42), after this table already had rows
    # from earlier days — those existing rows have no way to backfill the lap
    # they were predicted for, so they stay NULL permanently. Populated going
    # forward by prediction_worker._persist_and_publish from the same
    # lap-completion context that drives the rest of the prediction. See
    # strategy_service.get_strategy_prediction_history for the NULLS LAST
    # ordering this requires.
    lap_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # pit_predictor's own quick classifier-driven estimate — lap_number +
    # predicted_life_remaining (a genuine laps-until-degradation-threshold
    # count; was previously the raw tire_deg lap-time-delta prediction, a
    # bug fixed as part of the core-feature-rebuild's Checkpoint 4 — see
    # CLAUDE.md's now-closed "StrategyPrediction.optimal_pit_lap" Deferred
    # Wiring entry). Deliberately kept alongside recommended_pit_lap below,
    # not replaced by it — this is a different, complementary mechanism
    # (pit_predictor's degradation-threshold read vs. tire_deg's full
    # counterfactual stint-1/stint-2 search), not a duplicate.
    optimal_pit_lap: Mapped[int] = mapped_column(Integer, nullable=False)
    pit_probability: Mapped[float] = mapped_column(Float, nullable=False)
    undercut_score: Mapped[float] = mapped_column(Float, nullable=False)
    overcut_score: Mapped[float] = mapped_column(Float, nullable=False)
    # Laps remaining before predicted tyre-degradation crosses tire_deg_model.
    # DEGRADATION_THRESHOLD_SECONDS (tire_deg_model.predict_life_remaining_
    # batch's own output) — was previously the raw tire_deg lap-time-delta
    # prediction itself (a small, sometimes-negative float with no
    # "remaining laps" meaning at all despite the column name), a bug fixed
    # alongside optimal_pit_lap's own fix above (Checkpoint 4 — see
    # CLAUDE.md's now-closed "StrategyPrediction.tire_life_remaining"
    # Deferred Wiring entry).
    tire_life_remaining: Mapped[float] = mapped_column(Float, nullable=False)
    # P(the recommended_pit_lap candidate below is still optimal under
    # noise) — build_pit_recommendation's vectorized Monte Carlo. Was
    # previously always hardcoded 0.0 (never computed at all); real values
    # only exist from Checkpoint 4 onward — a NULL-free but meaningless 0.0
    # on any row predating this fix, same "no way to distinguish real-zero
    # from never-computed" caveat lap_number's own pre-migration NULL rows
    # already carry for a different column.
    confidence_score: Mapped[float] = mapped_column(Float, nullable=False)
    model_version: Mapped[str] = mapped_column(String(50), nullable=False)
    # --- The 5 columns below are new as of the core-feature-rebuild's
    # Checkpoint 4 migration (20260904_add_pit_recommendation_fields_to_
    # strategy_predictions) — strategy_service.build_pit_recommendation's
    # own output (Checkpoint 2: batched tire_deg counterfactual search over
    # stint-1/stint-2 candidates) plus get_pit_window_with_explanation's
    # combined tire_deg+pit_predictor explanation (Checkpoint 3), persisted
    # here so the live/replay per-lap pipeline (prediction_worker.
    # _persist_and_publish) carries the SAME rich recommendation the
    # on-demand /pit-window REST endpoint already computes, not just
    # pit_predictor's older, cruder optimal_pit_lap/pit_probability above.
    # All nullable: a row predating this migration has none of them, and a
    # row where build_pit_recommendation's own candidate search returns
    # nothing (e.g. current_lap >= total_laps, race essentially over) or any
    # sub-computation raises degrades to NULL here too — see
    # prediction_worker._compute_recommendation_fields's own docstring.
    recommended_pit_lap: Mapped[int | None] = mapped_column(Integer, nullable=True)
    window_start: Mapped[int | None] = mapped_column(Integer, nullable=True)
    window_end: Mapped[int | None] = mapped_column(Integer, nullable=True)
    recommended_compound: Mapped[str | None] = mapped_column(String(20), nullable=True)
    # Schemas.strategy_schema.PitRecommendationExplanation, serialized via
    # .model_dump(mode="json") — facts/narrative/tire_deg_shap/pit_predictor_shap.
    explanation: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

    session: Mapped["Session"] = relationship(back_populates="strategy_predictions")
    driver: Mapped["Driver"] = relationship(back_populates="strategy_predictions")


class PitEvent(Base):
    __tablename__ = "pit_events"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sessions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    driver_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("drivers.id"), nullable=False, index=True
    )
    lap_number: Mapped[int] = mapped_column(Integer, nullable=False)
    compound_in: Mapped[str] = mapped_column(String(10), nullable=False)
    compound_out: Mapped[str] = mapped_column(String(10), nullable=False)
    pit_duration_seconds: Mapped[float | None] = mapped_column(Float, nullable=True)
    was_predicted: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    session: Mapped["Session"] = relationship(back_populates="pit_events")
    driver: Mapped["Driver"] = relationship(back_populates="pit_events")

File: web/src/types/strategy.ts
// Mirrors backend/schemas/strategy_schema.py

export interface StrategyPredictionResponse {
  id: string
  session_id: string
  driver_id: string
  predicted_at: string
  optimal_pit_lap: number
  pit_probability: number
  undercut_score: number
  overcut_score: number
  tire_life_remaining: number
  confidence_score: number
  model_version: string
  created_at: string
}

// One SHAP feature contribution — see services/ml/explainability.py.
export interface FeatureContributionResponse {
  feature_name: string
  value: number
  contribution: number
  direction: string
}

// One structured fact behind a pit recommendation — see
// backend/schemas/strategy_schema.py's ExplanationFact.
export interface ExplanationFact {
  label: string
  value: string
  source: "tire_deg" | "pit_predictor" | "field"
}

// Combined tire_deg + pit_predictor explanation for one recommendation's #1
// candidate — see backend/schemas/strategy_schema.py's
// PitRecommendationExplanation. facts/narrative are server-derived from
// tire_deg_shap/pit_predictor_shap plus field/undercut context; both raw
// SHAP arrays are included too for a client that wants the full breakdown.
export interface PitRecommendationExplanation {
  facts: ExplanationFact[]
  narrative: string
  tire_deg_shap: FeatureContributionResponse[]
  pit_predictor_shap: FeatureContributionResponse[]
}

// window_start/window_end are identical across every candidate in one
// response — a narrow band around the #1 (rank 0) candidate (see
// build_pit_recommendation's own docstring), not a per-candidate range.
// confidence_score/explanation are populated only on the #1 candidate.
export interface PitWindowResponse {
  pit_lap: number
  window_start: number
  window_end: number
  projected_total_delta_seconds: number
  recommended_compound: string
  confidence_score: number | null
  explanation: PitRecommendationExplanation | null
}

export interface UndercutThreatResponse {
  target_driver_id: string
  probability_pit_now_gains_position: number
  projected_gap_seconds: number
  n_laps_projected: number
  recommended_action: string
}

export interface CompetitorStrategyEntry {
  driver_id: string
  predicted_pit_lap: number
  pit_probability: number
}

export interface StrategyOverviewResponse {
  session_id: string
  drivers: CompetitorStrategyEntry[]
}

export interface StrategyComparisonEntry {
  rank: number
  predicted_finishing_position: number
  strategy: StrategyPredictionResponse
}

export interface StrategyComparisonResponse {
  session_id: string
  driver_id: string
  strategies: StrategyComparisonEntry[]
}

// One StrategyPrediction row in a driver's lap-by-lap progression history —
// supplementary to StrategyOverviewResponse/UndercutThreatResponse (always
// live/current), used to reconstruct "the prediction valid at lap N" during
// replay/live progression. lap_number is null for pre-Day-42 rows (see
// backend docstring). recommended_pit_lap/window_start/window_end/
// recommended_compound/confidence_score/explanation are the core-feature-
// rebuild's Checkpoint 4 addition — the SAME rich recommendation
// PitWindowResponse carries, persisted per lap by prediction_worker so a
// replayed/live-progressing driver's history carries it too, not just
// pit_predictor's older, cruder predicted_pit_lap/pit_probability. All
// nullable except confidence_score (0.0 default) — a row predating that
// migration, or one where the computation degraded gracefully that lap, has
// none of them.
export interface StrategyPredictionHistoryEntry {
  lap_number: number | null
  predicted_pit_lap: number
  pit_probability: number
  undercut_score: number
  overcut_score: number
  created_at: string
  recommended_pit_lap: number | null
  window_start: number | null
  window_end: number | null
  recommended_compound: string | null
  confidence_score: number
  explanation: PitRecommendationExplanation | null
}

export interface StrategyPredictionHistoryResponse {
  session_id: string
  driver_id: string
  predictions: StrategyPredictionHistoryEntry[]
}

// GET /strategy/last-ingested-session — the R session with the newest
// race_date that has ingested lap data. The Strategy Simulator's session
// source when no race is live; resolved per-environment. event_name is null
// for rows ingested before that column existed (pre-2026) — fall back to
// circuit_name.
export interface LastIngestedSessionResponse {
  session_id: string
  season: number
  round_number: number
  event_name: string | null
  circuit_name: string
  race_date: string
}
